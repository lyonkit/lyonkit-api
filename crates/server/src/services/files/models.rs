use std::collections::HashMap;

use actix_web::{body::BoxBody, HttpRequest, HttpResponse, Responder};
use chrono::{DateTime, Utc};
use entity::file::Model;
use getset::Getters;
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Deserialize, Clone, Getters)]
#[serde(rename_all = "camelCase")]
#[getset(get = "pub")]
pub struct FileInput {
    pub content_type: Option<String>,
    pub content_length: u32,
    pub file_name: String,
    pub tags: Vec<String>,
    pub metadata: HashMap<String, String>,
}

#[derive(Serialize, Clone, Getters)]
#[serde(rename_all = "camelCase")]
#[getset(get = "pub")]
pub struct UploadFileOutput {
    pub id: i32,
    pub upload_url: String,
    pub key: String,
    pub public_url: String,
    pub tags: Vec<String>,
    pub metadata: Value,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl Responder for UploadFileOutput {
    type Body = BoxBody;

    fn respond_to(self, _req: &HttpRequest) -> HttpResponse<Self::Body> {
        HttpResponse::Ok().json(self)
    }
}

#[derive(Serialize, Clone, Getters)]
#[serde(rename_all = "camelCase")]
#[getset(get = "pub")]
pub struct FileOutput {
    pub id: i32,
    pub key: String,
    pub public_url: String,
    pub tags: Vec<String>,
    pub metadata: Value,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

pub struct FileOutputList(Vec<FileOutput>);

impl Responder for FileOutputList {
    type Body = BoxBody;

    fn respond_to(self, _req: &HttpRequest) -> HttpResponse<Self::Body> {
        HttpResponse::Ok().json(self.0)
    }
}

pub trait IntoFileOutputList {
    fn into_file_output_list(self, base_url: &str) -> FileOutputList;
}

impl IntoFileOutputList for Vec<Model> {
    fn into_file_output_list(self, base_url: &str) -> FileOutputList {
        let files = self
            .iter()
            .map(|model| {
                let public_url = format!("{}/{}", base_url, model.storage_key());
                FileOutput {
                    id: *model.id(),
                    public_url,
                    key: model.storage_key().clone(),
                    tags: model.tags().to_owned(),
                    metadata: model.metadata().clone(),
                    created_at: *model.created_at(),
                    updated_at: *model.updated_at(),
                }
            })
            .collect();

        FileOutputList(files)
    }
}

#[derive(Debug, Deserialize, Getters)]
#[getset(get = "pub")]
pub struct FileFilter {
    tag: Option<String>,
}
