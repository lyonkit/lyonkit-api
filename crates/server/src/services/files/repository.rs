use std::collections::HashMap;

use async_trait::async_trait;
use sea_orm::{prelude::*, sea_query::Expr, ActiveModelTrait, ActiveValue::Set, ConnectionTrait};
use serde_json::{Map, Value};
use uuid::Uuid;

use entity::file::{ActiveModel, Column, Entity, Model};

use crate::{
    errors::{utils::MapApiError, ApiError},
    services::files::models::{FileInput, FileUpdateInput},
};

#[async_trait]
pub trait FilesRepository {
    async fn create_file(&self, namespace: &str, file: FileInput) -> Result<Model, ApiError>;
    async fn find_all_by_tag(
        &self,
        namespace: &str,
        tag: &Option<String>,
    ) -> Result<Vec<Model>, ApiError>;
    async fn update_file(
        &self,
        namespace: &str,
        id: &i32,
        payload: FileUpdateInput,
    ) -> Result<Model, ApiError>;
    async fn delete_file(&self, namespace: &str, id: &i32) -> Result<Model, ApiError>;
}

fn string_map_to_json(map: &HashMap<String, String>) -> Value {
    let mut value = Map::new();

    for (key, str_value) in map {
        value.insert(key.to_string(), Value::String(str_value.to_owned()));
    }

    Value::Object(value)
}

#[async_trait]
impl<T: ConnectionTrait> FilesRepository for T {
    async fn create_file(&self, namespace: &str, input: FileInput) -> Result<Model, ApiError> {
        let storage_key = format!(
            "{}_{}",
            Uuid::new_v4().to_string().replace('-', ""),
            input.file().file_name()
        );
        let file_model = entity::file::ActiveModel {
            namespace: Set(namespace.to_owned()),
            storage_key: Set(storage_key),
            tags: Set(input.tags().to_owned()),
            metadata: Set(string_map_to_json(input.metadata())),
            ..Default::default()
        };

        let inserted_file = file_model.insert(self).await.map_api_err()?;
        Ok(inserted_file)
    }

    async fn find_all_by_tag(
        &self,
        namespace: &str,
        tag: &Option<String>,
    ) -> Result<Vec<Model>, ApiError> {
        let mut query = Entity::find().filter(Column::Namespace.eq(namespace.to_owned()));

        if let Some(t) = tag {
            query = query.filter(Expr::cust_with_values(
                r#"$1 = any(tags)"#,
                vec![t.to_string()],
            ));
        }

        let files = query.all(self).await.map_api_err()?;

        Ok(files)
    }

    async fn update_file(
        &self,
        namespace: &str,
        id: &i32,
        payload: FileUpdateInput,
    ) -> Result<Model, ApiError> {
        let mut active_model: ActiveModel = Entity::find()
            .filter(Column::Namespace.eq(namespace.to_owned()))
            .filter(Column::Id.eq(id.to_owned()))
            .one(self)
            .await?
            .ok_or(ApiError::NotFound)?
            .into();

        if let Some(tags) = payload.tags() {
            active_model.tags = Set(tags.to_owned());
        }

        if let Some(metadata) = payload.metadata() {
            active_model.metadata = Set(string_map_to_json(metadata));
        }

        if let Some(file) = payload.file() {
            let storage_key = format!(
                "{}_{}",
                Uuid::new_v4().to_string().replace('-', ""),
                file.file_name()
            );
            active_model.storage_key = Set(storage_key);
        }

        let model = active_model.update(self).await?;

        Ok(model)
    }

    async fn delete_file(&self, namespace: &str, id: &i32) -> Result<Model, ApiError> {
        let model = Entity::find()
            .filter(Column::Namespace.eq(namespace.to_owned()))
            .filter(Column::Id.eq(id.to_owned()))
            .one(self)
            .await?
            .ok_or(ApiError::NotFound)?;

        model.clone().delete(self).await?;
        Ok(model)
    }
}
