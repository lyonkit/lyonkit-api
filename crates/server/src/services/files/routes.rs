use std::time::Duration;

use actix_web::{get, post, web};
use aws_sdk_s3::{model::ObjectCannedAcl::PublicRead, presigning::config::PresigningConfig};

use crate::{
    errors::ApiError,
    middlewares::{
        api_key::{ApiKey, WriteApiKey},
        s3::S3ClientProvider,
    },
    server::AppState,
    services::files::{
        models::{FileFilter, FileInput, FileOutputList, IntoFileOutputList, UploadFileOutput},
        repository::FilesRepository,
    },
};

#[post("")]
pub async fn create_file(
    data: web::Data<AppState>,
    api_key: WriteApiKey,
    s3_provider: S3ClientProvider,
    file_data: web::Json<FileInput>,
) -> Result<UploadFileOutput, ApiError> {
    let maybe_content_type = file_data.content_type.clone();
    let content_length = *file_data.content_length();
    let max_file_size = 50_000_000_u32;
    if content_length > max_file_size {
        return Err(ApiError::FileTooBig(max_file_size, content_length));
    }

    let model = data
        .conn()
        .create_file(api_key.namespace().as_str(), file_data.into_inner())
        .await?;

    let settings = data.settings();
    let s3_client = s3_provider.provide();

    let mut s3_request = s3_client
        .put_object()
        .content_length(content_length.into())
        .bucket(settings.s3().buckets().file())
        .key(model.storage_key())
        .acl(PublicRead);

    if let Some(content_type) = maybe_content_type {
        s3_request = s3_request.content_type(content_type)
    }

    let presigned_url = s3_request
        .presigned(
            PresigningConfig::builder()
                .expires_in(Duration::from_secs(180))
                .build()
                .unwrap(),
        )
        .await
        .map_err(|_e| ApiError::InternalServerError)?
        .uri()
        .to_string();

    let public_url = format!(
        "{}/{}/{}",
        settings.s3().base_url(),
        settings.s3().buckets().file(),
        model.storage_key()
    );
    Ok(UploadFileOutput {
        id: model.id,
        upload_url: presigned_url,
        key: model.storage_key,
        tags: model.tags,
        metadata: model.metadata,
        public_url,
        created_at: model.created_at,
        updated_at: model.updated_at,
    })
}

#[get("")]
pub async fn list_files(
    data: web::Data<AppState>,
    api_key: ApiKey,
    filter: web::Query<FileFilter>,
) -> Result<FileOutputList, ApiError> {
    let files = data
        .conn()
        .find_all_by_tag(api_key.namespace().as_str(), filter.tag())
        .await?;

    let s3_settings = data.settings().s3();
    let s3_base_url = format!(
        "{}/{}",
        s3_settings.base_url(),
        s3_settings.buckets().file()
    );

    Ok(files.into_file_output_list(s3_base_url.as_str()))
}
