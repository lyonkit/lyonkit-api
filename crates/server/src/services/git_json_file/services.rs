use crate::errors::ApiError;
use crate::services::git_json_file::models::GitJsonFile;
use base64::decode as decode_b64;
use lazy_static::lazy_static;
use reqwest::Client;
use tracing::error;

lazy_static! {
  pub static ref GITHUB_CLIENT: reqwest::Client =
    Client::builder().https_only(true).build().unwrap();
}

pub(crate) async fn fetch_git_json_file(
  path: &String,
  namespace: &String,
  org: &String,
  repo: &String,
  github_token: &String,
) -> Result<GitJsonFile, ApiError> {
  let url = format!("https://api.github.com/repos/{org}/{repo}/contents/{path}");
  let response = GITHUB_CLIENT
    .get(&url)
    .header("Authorization", github_token)
    .header("User-Agent", format!("LyonkitApi ({})", namespace))
    .send()
    .await
    .map_err(|err| {
      error!(
        url = &url,
        error = format!("{:?}", &err),
        "An error occured while querying git file"
      );
      ApiError::GitError
    })?;

  // On decode le body de b64 à String à JSON
  let mut body = response.json::<GitJsonFile>().await.map_err(|err| {
    error!(
      url = &url,
      error = format!("{:?}", &err),
      "Failed to parse body of retrieved git file"
    );
    ApiError::GitBodyUnparseable
  })?;

  let content = body
    .content()
    .as_str()
    .map(|c| c.to_string())
    .ok_or_else(|| "no content field".to_string())
    .and_then(|b64_content| {
      decode_b64(b64_content.replace('\n', "")).map_err(|err| format!("{:?}", err))
    })
    .and_then(|bytes| {
      serde_json::from_slice::<serde_json::Value>(bytes.as_slice())
        .map_err(|err| format!("{:?}", err))
    })
    .map_err(|err| {
      error!(
        url = &url,
        error = format!("{:?}", &err),
        "Failed to parse body of retrieved git file"
      );
      ApiError::GitBodyUnparseable
    })?;

  body.set_content(content);

  Ok(body)
}
