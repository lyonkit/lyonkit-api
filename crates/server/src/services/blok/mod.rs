use crate::services::blok::routes::{create_blok, delete_blok, get_blok, patch_blok, update_blok};
use actix_web::{web::scope, Scope};

pub(crate) mod models;
mod routes;

pub fn blok_service() -> Scope {
    scope("/blok")
        .service(get_blok)
        .service(create_blok)
        .service(update_blok)
        .service(patch_blok)
        .service(delete_blok)
}
