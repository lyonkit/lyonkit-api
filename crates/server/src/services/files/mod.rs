use actix_web::{web::scope, Scope};

use crate::services::files::routes::{create_file, delete_file, list_files, update_file};

pub mod models;
pub mod repository;
pub mod routes;

pub fn file_service() -> Scope {
    scope("/file")
        .service(create_file)
        .service(list_files)
        .service(update_file)
        .service(delete_file)
}
