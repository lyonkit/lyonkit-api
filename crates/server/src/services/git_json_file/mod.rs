mod models;
mod routes;
mod services;

use crate::services::git_json_file::routes::{get_git_json_file, update_git_json_file};
use actix_web::{web::scope, Scope};

pub fn git_json_file_service() -> Scope {
    scope("/git/json-file")
        .service(get_git_json_file)
        .service(update_git_json_file)
}
