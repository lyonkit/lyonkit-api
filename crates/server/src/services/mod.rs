mod blok;
mod image;
mod git_json_file;
mod page;
mod post;
mod quote;

use crate::middlewares::s3::S3ProviderMiddlewareFactory;
use crate::services::image::image_service;
use crate::services::git_json_file::{git_json_file_service};
use crate::services::post::post_service;
use crate::services::quote::quote_service;
use crate::{
  middlewares::api_key::ApiKeyMiddlewareFactory,
  services::{blok::blok_service, page::page_service},
};
use actix_web::{
  dev::{ServiceFactory, ServiceRequest, ServiceResponse},
  get, web, HttpResponse, Responder, Scope,
};
use serde_json::json;

#[get("/ping")]
async fn ping() -> impl Responder {
  HttpResponse::Ok().json(json!({
    "message": "Hello LyonKit API !"
  }))
}

pub fn api_services() -> Scope<
  impl ServiceFactory<
    ServiceRequest,
    Response = ServiceResponse,
    Error = actix_web::Error,
    Config = (),
    InitError = (),
  >,
> {
  web::scope("/api")
    .app_data(web::JsonConfig::default().error_handler(|err, _req| {
      actix_web::error::InternalError::from_response(
        "",
        HttpResponse::BadRequest()
          .content_type("application/json")
          .body(format!(r#"{{"code": "JSNER", "message":"{}"}}"#, err)),
      )
      .into()
    }))
    .wrap(ApiKeyMiddlewareFactory::new())
    .wrap(S3ProviderMiddlewareFactory::new())
    .service(ping)
    .service(page_service())
    .service(blok_service())
    .service(image_service())
    .service(post_service())
    .service(quote_service())
    .service(git_json_file_service())
}
