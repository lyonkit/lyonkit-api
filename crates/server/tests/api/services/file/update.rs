use crate::test_app::TestApp;
use insta::assert_json_snapshot;
use reqwest::StatusCode;
use serde_json::json;
use server::services::files::{
    models::{FileInput, FilePayload},
    repository::FilesRepository,
};
use std::collections::HashMap;
use test_context::test_context;

#[test_context(TestApp)]
#[tokio::test]
async fn update_file_metadata_should_work(ctx: &mut TestApp) {
    ctx.create_api_key("test_update_file", false).await;

    let model = ctx
        .database_connection()
        .create_file(
            "test_update_file",
            FileInput {
                file: FilePayload {
                    file_name: "test_file.txt".to_string(),
                    content_type: Some("text/plain".to_string()),
                    content_length: 100000,
                },
                metadata: HashMap::from([("update".to_string(), "no".to_string())]),
                tags: vec!["events".to_string()],
            },
        )
        .await
        .expect("Failed to create test file");

    let id = model.id();

    let response = ctx
        .put(
            format!("/file/{id}"),
            json!({
                "metadata": {
                    "update": "yes"
                }
            }),
        )
        .await;

    assert_eq!(StatusCode::OK, response.status());

    let body = response
        .json::<serde_json::Value>()
        .await
        .expect("Failed to deserialize body");

    assert!(body.get("id").is_some());
    assert!(body.get("createdAt").is_some());
    assert!(body.get("updatedAt").is_some());
    assert!(body.get("publicUrl").is_some());
    assert!(body.get("key").is_some());
    assert_eq!(Some(&serde_json::Value::Null), body.get("uploadUrl"));

    let mut body = (*body.as_object().expect("Body is not a JSON object")).clone();

    // Remove unpredictable values for snapshot
    body.remove("id");
    body.remove("createdAt");
    body.remove("updatedAt");
    body.remove("publicUrl");
    body.remove("key");

    assert_json_snapshot!(body);
}

#[test_context(TestApp)]
#[tokio::test]
async fn update_file_should_work(ctx: &mut TestApp) {
    ctx.create_api_key("test_update_file", false).await;

    let model = ctx
        .database_connection()
        .create_file(
            "test_update_file",
            FileInput {
                file: FilePayload {
                    file_name: "test_file.txt".to_string(),
                    content_type: Some("text/plain".to_string()),
                    content_length: 100000,
                },
                metadata: HashMap::from([("update".to_string(), "no".to_string())]),
                tags: vec!["events".to_string()],
            },
        )
        .await
        .expect("Failed to create test file");

    let id = model.id();

    let response = ctx
        .put(
            format!("/file/{id}"),
            json!({
                "file": {
                    "contentType": "text/plain",
                    "contentLength": 10000,
                    "fileName": "file.txt"
                }
            }),
        )
        .await;

    assert_eq!(StatusCode::OK, response.status());

    let body = response
        .json::<serde_json::Value>()
        .await
        .expect("Failed to deserialize body");

    assert!(body.get("id").is_some());
    assert!(body.get("createdAt").is_some());
    assert!(body.get("updatedAt").is_some());
    assert!(body.get("publicUrl").is_some());
    assert!(body.get("key").is_some());
    assert!(body.get("uploadUrl").is_some());

    let mut body = (*body.as_object().expect("Body is not a JSON object")).clone();

    // Remove unpredictable values for snapshot
    body.remove("id");
    body.remove("createdAt");
    body.remove("updatedAt");
    body.remove("publicUrl");
    body.remove("key");
    body.remove("uploadUrl");

    assert_json_snapshot!(body);
}
