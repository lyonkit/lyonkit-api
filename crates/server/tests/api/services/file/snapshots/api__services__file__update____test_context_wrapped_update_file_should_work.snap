---
source: crates/server/tests/api/services/file/update.rs
expression: body
---
{
  "metadata": {
    "update": "no"
  },
  "tags": [
    "events"
  ]
}
